cli/class-block-plugin-checker.php

```diff
@@ -0,0 +1,476 @@
+ * (particularly the prepare*() and find*() functions), and a separate class for the checks.
+ *
+ * @package WordPressdotorg\Plugin_Directory\CLI
+ */
+class Block_Plugin_Checker {
+
+	protected $allowed_hosts = array(
+		'github.com',
+		'plugins.svn.wordpress.org',
+	);
+
+	protected $path_to_plugin = null;
+	protected $check_methods = array();
+	protected $results = array();
+
+	public $slug = null;
+	public $repo_url = null;
+	public $repo_revision = null;
+	protected $readme_path = null;
+	protected $readme = null;
+	protected $headers = null;
+	protected $blocks = null;
+	protected $block_json_files = null;
+	protected $block_json = array();
+
+	/**
+	 * Constructor.
+	 *
+	 * @param string $slug The plugin slug, if known. Optional.
+	 */
+	public function __construct( $slug = null ) {
+		$this->slug = $slug;
+	}
+
+	/**
+	 * Check a plugin given the URL of a Subversion or GitHub repo.
+	 * Note that only hosts listed in $allowed_hosts are permitted.
+	 *
+	 * @param string $url The URL of a Subversion or GitHub repository.
+	 * @return array A list of status items.
+	 */
+	public function run_check_plugin_repo( $url ) {
+
+		// git@github.com:sortabrilliant/jumbotron.git
+
+		if ( preg_match( '#^\w+@github[.]com:(\w+/[^.]+)\.git$#', $url, $matches ) ) {
+			$url = 'https://github.com/' . $matches[1];
+		}
+
+		// Parse the URL with whitespace and trailing / trimmed
+		$url_parts = wp_parse_url( rtrim( trim( $url ), '/' ) );
+
+		if ( empty( $url_parts ) ) {
+			$this->record_result(
+				__FUNCTION__,
+				'error',
+				sprintf( __( 'Invalid url %s' ), $url ),
+				$url
+			);
+			return $this->results;
+		}
+
+		if ( empty( $url_parts['host'] ) || !in_array( $url_parts['host'], $this->allowed_hosts ) ) {
+			$this->record_result(
+				__FUNCTION__,
+				'error',
+				sprintf( __( 'URL must be GitHub or plugins.svn.wordpress.org %s' ), $url ),
+				$url
+			);
+			return $this->results;
+		}
+
+		if ( 'plugins.svn.wordpress.org' === $url_parts['host'] ) {
+			$path_parts = explode( '/', $url_parts[ 'path' ], 3 );
+			if ( empty( $path_parts[1] ) ) {
+				$this->record_result(
+					__FUNCTION__,
+					'error',
+					sprintf( __( 'URL must be a plugin repository %s' ), $url ),
+					$url
+				);
+				return $this->results;
+			}
+
+			$this->slug = $path_parts[1];
+			$url = 'https://plugins.svn.wordpress.org/' . $path_parts[1];
+			if ( 4 === count( $path_parts )
+				&& ( 'tags' === $path_parts[2] || 'branches' === $path_parts[2] )
+				&& ( !empty( $path_parts[3] ) ) ) {
+				$url .= $path_parts[2] . '/' . $path_parts[3];
+			} else {
+				$url .= '/trunk';
+			}
+		} elseif ( 'github.com' === $url_parts['host'] ) {
+			// https://github.com/sortabrilliant/jumbotron
+			// git@github.com:sortabrilliant/jumbotron.git
+			// https://github.com/sortabrilliant/jumbotron.git
+			//
+			if ( preg_match( '#^/(\w+/[^.]+)\.git#', $url_parts[ 'path' ], $matches ) ) {
```

Comment:
There's a regex at line 59 that converts a `git@github.com` formatted URL. I figured it might as well work with any GitHub URL format someone could use to refer to a repo.

Reply:
I guess it's fair to handle any format.

My point however was that a URL **excluding .git$** would fail to pass the check, including a `git@..` url due to the regex on line 59.

Reply:
Oh, I see what you mean now! You're absolutely right, fixing that..

Reply:
Fixed in 2cd8d0e.